src/natcap/invest/coastal_blue_carbon/preprocessor.py

```diff
@@ -76,292 +67,175 @@ def execute(args):
+            args['landcover_snapshot_csv']))
+
+    # Align the raster stack for analyzing the various transitions.
+    min_pixel_size = float('inf')
+    source_snapshot_paths = []
+    aligned_snapshot_paths = []
+    for snapshot_year, raster_path in snapshots_dict.items():
+        source_snapshot_paths.append(raster_path)
+        aligned_snapshot_paths.append(os.path.join(
+            output_dir, ALIGNED_LULC_RASTER_TEMPLATE.format(
+                year=snapshot_year, suffix=suffix)))
+        min_pixel_size = min(
+            utils.mean_pixel_size_and_area(
+                pygeoprocessing.get_raster_info(raster_path)['pixel_size'])[0],
+            min_pixel_size)
+
+    baseline_srs_wkt = pygeoprocessing.get_raster_info(
+        snapshots_dict[min(snapshots_dict.keys())])['projection_wkt']
+    alignment_task = task_graph.add_task(
+        func=pygeoprocessing.align_and_resize_raster_stack,
+        args=(source_snapshot_paths,
+              aligned_snapshot_paths,
+              (['near']*len(source_snapshot_paths)),
+              (min_pixel_size, -min_pixel_size),
+              'intersection'),
+        kwargs={'target_projection_wkt': baseline_srs_wkt},
+        hash_algorithm='md5',
+        copy_duplicate_artifact=True,
+        target_path_list=aligned_snapshot_paths,
+        task_name='Align input landcover rasters')
+
+    landcover_table = utils.build_lookup_from_csv(
+        args['lulc_lookup_table_path'], 'code')
+
+    target_transition_table = os.path.join(
+        output_dir, TRANSITION_TABLE.format(suffix=suffix))
+    _ = task_graph.add_task(
+        func=_create_transition_table,
+        args=(landcover_table,
+              sorted(snapshots_dict.values(), key=lambda x: [0]),
+              target_transition_table),
+        target_path_list=[target_transition_table],
+        dependent_task_list=[alignment_task],
+        task_name='Determine transitions and write transition table')
+
+    # Creating this table should be cheap, so it's not in a task.
+    # This is only likely to be expensive if the user provided a landcover
+    # lookup with many many rows. If that happens, this tool will have other
+    # problems (like exploding memory in the transition table creation).
+    target_biophysical_table_path = os.path.join(
+        output_dir, BIOPHYSICAL_TABLE.format(suffix=suffix))
+    _create_biophysical_table(landcover_table, target_biophysical_table_path)
+
+    task_graph.close()
+    task_graph.join()
+
+
+def _create_transition_table(landcover_table, lulc_snapshot_list,
+                             target_table_path):
+
+    def _read_block(raster_path, offset_dict):
+        try:
+            raster = gdal.OpenEx(raster_path, gdal.OF_RASTER)
+            band = raster.GetRasterBand(1)
+            array = band.ReadAsArray(**offset_dict)
+            nodata = band.GetNoDataValue()
+        finally:
+            band = None
+            raster = None
+        return array, nodata
```

Comment:
This is a fun function that I feel like you could use in the main module where I left a comment about opening a raster file per iterblock. If that is something you want to do, this could be a nice function to simplify that code in the other file.

Reply:
I've been profiling the reimplementation, and found that out of a total profiled runtime of 1338.949s, only about 10.5s (about 0.8% of total runtime) was spent across almost 6500 calls to `gdal.OpenEx`:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     6776   10.498    0.002   10.498    0.002 {built-in method _gdal.OpenEx}
```

So I feel comfortable with just opening the raster when needed, reading the block and closing the raster, and since there may well be a [practical limit to the number of open files allowed](https://docs.microsoft.com/en-us/cpp/c-runtime-library/file-handling?redirectedfrom=MSDN&view=vs-2019), I think this would be safest.  I'll patch this shortly.

Reply:
Patched in f118a2b2 so that both the model and the preprocessor use a shared function that reads a block on demand.

Reply:
@richpsharp just wanted to let you know that I've settled on keeping the raster objects in a list.  After some further profiling, it became clear that I was reading my initial profile results incorrectly and that replacing these repeated `gdal.OpenEx`s with using the original raster values, I could trim the raster summing function's runtime in the main model by about 40%, which has been done (in both the model and preprocessor) in 84a2a20c.